### PR TITLE
Python Wrapper: LenientNamedTuple

### LINKED ISSUE #6999: Python Wrapper: LenientNamedTuple
Create a lenient implementation of NamedTuple which ignores unknown fields.
Will allow the implementation of SDK data structure to be future compatible (won't break the wrapper)

### PR DESCRIPTION
**Pull request by N-o-Z**
Closes #6999

## Change Description

Suggest a solution for an immutable namedtuple that can ignore unknown fields

### Background

Ability to support SDK object modeling with future compatibility guarantees 

### Testing Details

Unit and System tested

### CODE REVIEW

**Comment by N-o-Z** [comment]
One reviewer is enough

File: clients/python-wrapper/tests/utests/test_namedtuple.py [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+        "field1": "test",
+        "field2": 2,
+        "field3": False,
+        "field4": "something",
+        "field5": None,
+    }
+    nt = NamedTupleTest(**kwargs)
+    assert nt.field1 == "test"
+    assert nt.field2 == 2
+    assert not nt.field3
```
Comment (arielshaqed) [inline clients/python-wrapper/tests/utests/test_namedtuple.py:37]: Not sure what this tests.

Comment (arielshaqed) [inline clients/python-wrapper/tests/utests/test_namedtuple.py:37]: Can you add a test that `field4` and `field5` are _not_ present?  That would clarify behaviour.

**Review by arielshaqed — CHANGES_REQUESTED** [review]
Better, but...  This still ties together wrapper client versions with server versions!  A server with a newer version will be usable with an older wrapper client, but there is no way to read any added optional fields.  To resolve I need to update the wrapper version, updating merely the SDK version is not enough.

TBH I don't understand this requirement not to return `ObjectStats`.  @ozkatz / @N-o-Z can you please clarify?  Having 2 object classes with identical or nearly-identical APIs seems unpythonic, harder to use, and higher maintenance.  Not sure what the upside is to justify this.

File: clients/python-wrapper/tests/utests/test_namedtuple.py (hunk 1 above)
Comment (N-o-Z) [reply clients/python-wrapper/tests/utests/test_namedtuple.py:37]: Testing that we create the object successfully even when providing unknown arguments

Comment (N-o-Z) [reply clients/python-wrapper/tests/utests/test_namedtuple.py:37]: Done

**Comment by ozkatz** [comment]
Some context to my objection on exposing lakefs_sdk members:
1. abstraction leak - the `lakefs` library currently uses lakefs_sdk for *most* but not *all* calls to remote server. I see it as an implementation detail. Perhaps at some point we want to remove that dependency and use urllib? This leaks that implementation detail and increases surface area as most types also come with methods such as `to_json()`, `to_dict()` and `from_orm()`.
3. ergonomics - 99% of `lakefs_sdk` is generated code, documented in a different place, with generated documentation as well. To make sense of the `lakefs` library, users have to jump between 2 documentation sites, one of which we know is not great.

That being said, this is an opinion and not a requirement, especially since concern 1 is purely an engineering one.

**Comment by N-o-Z** [comment]
@arielshaqed  I feel the same as @ozkatz and suggest we continue with the PR review

File: clients/python-wrapper/tests/utests/test_namedtuple.py [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+    try:
+        nt.field4
+        assert 0, "Expected exception"
```
Comment (arielshaqed) [inline clients/python-wrapper/tests/utests/test_namedtuple.py:40]: Consider
```suggestion
        raise AssertionError("Expected AttributeError")
```
or using the `unittest` module.

**Review by arielshaqed — APPROVED** [review]
Thanks!

File: clients/python-wrapper/tests/utests/test_namedtuple.py (hunk 2 above)
Comment (N-o-Z) [reply clients/python-wrapper/tests/utests/test_namedtuple.py:40]: I have a context for that - I'll replace all occurrences

File: clients/python-wrapper/lakefs/namedtuple.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+    Class which provides the NamedTuple functionality but allows initializing with unknown fields.
+    The unknown fields will be ignored and mandatory fields are enforced
+    """
+
+    def __new__(cls, **kwargs):
+        fields = list(cls.__dict__["__annotations__"].keys())
+        for k, v in kwargs.items():
+            if k in fields:
+                setattr(cls, k, v)
+                fields.remove(k)
```
[the remark this answers is not in the document]
Comment (N-o-Z) [reply clients/python-wrapper/lakefs/namedtuple.py:17]: Great idea! Will implement

**Pull request merged**